### PR TITLE
Anomaly metric

### LINKED ISSUE #1841: AnomalyMetric for optimization of anomaly params during swarming (expected anomaly %)
Currently, it is possible to swarm over various anomaly settings (`slidingWindowSize, mode, `) but it is actually pointless, as even the `TemporalAnomaly` model type optimizes for best prediction, the anomaly scores are nowhere taken to account in metric computation. 

I suggest tackling this by a new metric that would allow you to say "in this dataset, I expect anomaly to occur in about 1% of samples": `AnomalyMetric(windowSize=1000, expectedOccurence=0.01)` 

This would: 
* give a meaning to the anomaly params during swarming
* allow user to adjust how much anomalies they detect (sensitivity) in their data -- this is a common issue

### PR DESCRIPTION
**Pull request by breznak**
Fixes #1841 

creates a metric that can optimize swarming parameters by `anomaly score`

Note: this PR is based on (unmerged) #1929

### CODE REVIEW

**Comment by breznak** [comment]
@scottpurdy @oxtopus @subutai please for your early review. 
This PR is ready, but blocked by how we resolve #1929 as it builds atop of it. For your review, you can use the following view: https://github.com/breznak/nupic/compare/metric_expose_model...breznak:anomaly_metric

File: examples/swarm/anomaly_swarming_hotgym/swarm/description.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Template file used by the OPF Experiment Generator to generate the actual
+description.py file by replacing $XXXXXXXX tokens with desired values.
+
+This description.py file was generated by:
+'~/nupic/eng/lib/python2.6/site-packages/nupic/frameworks/opf/expGenerator/ExpGenerator.py'
```
Comment (oxtopus) [inline examples/swarm/anomaly_swarming_hotgym/swarm/description.py]: 2.6? :confused:

File: nupic/frameworks/opf/model.py [hunk 2]
```diff
@@ -24,12 +24,16 @@
...
 import shutil
+import logging
 from abc import ABCMeta, abstractmethod
 
 import nupic.frameworks.opf.opfutils as opfutils
 
-
 ###############################################################
+# global variable
+global globalModelsStorage
```
Comment (oxtopus) [inline nupic/frameworks/opf/model.py]: This is the wrong context for usage of `global` keyword.  It need only be used inside functions.

File: nupic/frameworks/opf/model.py [hunk 3]
```diff
@@ -126,6 +134,30 @@ def _getLogger(self):
...
+
+    newModel = self
+    logger = logging.getLogger() # self._getLogger() not working because of subclass init
+
+    # generate unique name
+    if newModel._name is None:
+      newModel._name = "model-"+str(id(newModel))
+
+    name = newModel._name
+    if name in globalModelsStorage.keys():
```
Comment (oxtopus) [inline nupic/frameworks/opf/model.py]: This need only be `if name in globalModelsStorage`, which is faster anyway.

File: examples/swarm/anomaly_swarming_hotgym/swarm/description.py (hunk 1 above)
Comment (breznak) [inline examples/swarm/anomaly_swarming_hotgym/swarm/description.py]: ah, I just copied the file from `examples/.....multistep..../hotgym`, can I just delete the line altogether?

Comment (oxtopus) [inline examples/swarm/anomaly_swarming_hotgym/swarm/description.py]: Sure, especially if it wasn't autogenerated :stuck_out_tongue_winking_eye:

File: nupic/frameworks/opf/model.py (hunk 2 above)
Comment (breznak) [inline nupic/frameworks/opf/model.py]: :+1: thanks, applying the fix to the original branch.

**Comment by rhyolight** [comment]
Status on this?

**Comment by breznak** [comment]
@rhyolight blocked-by #1842

File: examples/swarm/anomaly_swarming_hotgym/description.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+      MetricSpec(field=config['predictedField'], metric='trivial',
+                 inferenceElement='prediction',
+                 params={'errorMetric': 'altMAPE', 'window': 1000, 'steps': steps}))
+  control['metrics'].append(
+      MetricSpec(field=config['predictedField'], metric='anomaly',
+                 inferenceElement='prediction',
+                 params={'errorMetric': 'altMAPE', 'window': 100, 'steps': steps,
+                         'desiredPct': 0.1, 'modelName': config['modelParams']['name']}))
+
+################################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/description.py]: Just remove these smudged lines and keep a space of 3 line breaks after method, please.

File: examples/swarm/anomaly_swarming_hotgym/permutations.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# Permutation optimization setting: either minimize or maximize metric
+# used by RunPermutations.
+# NOTE: The value is used as a regular expressions by RunPermutations.py's
+#       report generator
+# (generated from minimize = 'prediction:aae:window=1000:field=consumption')
+#minimize = "multiStepBestPredictions:multiStep:errorMetric='aae':steps=1:window=1000:field=consumption"
+minimize = "prediction:anomaly:desiredPct=0.1:errorMetric='altMAPE':modelName='hotgymAnomalySwarmingDemo':steps=1:window=100:field=consumption"
+
+
+#############################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/permutations.py]: Please remove these smudged lines.

File: examples/swarm/anomaly_swarming_hotgym/swarm/description.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+      MetricSpec(field=config['predictedField'], metric='trivial',
+                 inferenceElement='prediction',
+                 params={'errorMetric': 'altMAPE', 'window': 1000, 'steps': steps}))
+  control['metrics'].append(
+      MetricSpec(field=config['predictedField'], metric='anomaly',
+                 inferenceElement='multiStepBestPredictions',
+                 params={'errorMetric': 'altMAPE', 'window': 1000, 'steps': steps,
+                         'desiredPct': 0.1, 'modelName': config['name']}))
+
+################################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/swarm/description.py]: Please remove these smudged lines and keep a space of 3 line breaks after previous method.

File: examples/swarm/anomaly_swarming_hotgym/swarm/permutations.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+# Permutation optimization setting: either minimize or maximize metric
+# used by RunPermutations.
+# NOTE: The value is used as a regular expressions by RunPermutations.py's
+#       report generator
+# (generated from minimize = 'prediction:aae:window=1000:field=consumption')
+minimize = "multiStepBestPredictions:anomaly:desiredPct=0.1:errorMetric='altMAPE':modelName='hotgymAnomalySwarmingDemo':steps=1:window=1000:field=consumption"
+
+
+
+#############################################################################
```
Comment (david-ragazzi) [inline examples/swarm/anomaly_swarming_hotgym/swarm/permutations.py]: Please remove this smudged line aboved method.

File: nupic/frameworks/opf/clamodel.py [hunk 8]
```diff
@@ -1498,4 +1503,84 @@ def getNextRecordDict(self):
 
 
 
+###############################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/clamodel.py]: Please remove this smudged line.

File: nupic/frameworks/opf/metrics.py [hunk 9]
```diff
@@ -1482,3 +1486,96 @@ def __repr__(self):
   def getMetric(self):
     return {'value': self.err, "stats" : {"weights" : self.weights}}
 
+
+
+#############################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Please remove this smudged line.

File: nupic/frameworks/opf/metrics.py [hunk 10]
```diff
@@ -1482,3 +1486,96 @@ def __repr__(self):
...
+
+  
+  def getModelInstance(self):
+    """@return model linked to this metric"""
+    model = GlobalDict.get(self._activeModel)
+    if model is None:
+      raise ValueError("MetricModelCallback: failed to access model named '%s' " % (self._activeModel))
+    return model
+
+#################################################################################
```
Comment (david-ragazzi) [inline nupic/frameworks/opf/metrics.py]: Please remove this smudged line.

File: nupic/utils.py [hunk 11]
```diff
@@ -97,6 +98,63 @@ def __setstate__(self, state):
...
+        o.windowSize == self.windowSize):
+      return 0
+    else: 
+      return -1
+
+
   def __call__(self, value):
     return self.next(value)
 
+#######################################################################
```
Comment (david-ragazzi) [inline nupic/utils.py]: Please remove this smudged line.

**Comment by rhyolight** [comment]
This PR has been open for way, way too long. @breznak I'm sorry to close it because I know you put a lot of work into it, but maybe it is time to try again from a clean slate. So many changes here, and hasn't been synced with master in ages.

**Pull request closed without merging**